src/logics/magicsea/logic.swap-token.test.ts

```diff
@@ -0,0 +1,155 @@
+        options: { account: '0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa' },
+      },
+      {
+        chainId: common.ChainId.iota,
+        fields: {
+          tradeType: core.TradeType.exactOut,
+          input: new common.TokenAmount(iotaTokens.IOTA, '6'),
+          output: new common.TokenAmount(iotaTokens['USDC.e'], '1'),
+          path: [],
+          slippage: 500,
+        },
+        options: { account: '0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa' },
+      },
+      {
+        chainId: common.ChainId.iota,
+        fields: {
+          tradeType: core.TradeType.exactOut,
+          input: new common.TokenAmount(iotaTokens['USDC.e'], '1'),
+          output: new common.TokenAmount(iotaTokens.IOTA, '6'),
+          path: [],
+          slippage: 500,
+        },
+        options: { account: '0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa' },
+      },
+      {
+        chainId: common.ChainId.iota,
+        fields: {
+          tradeType: core.TradeType.exactOut,
+          input: new common.TokenAmount(iotaTokens['USDC.e'], '1'),
+          output: new common.TokenAmount(iotaTokens.GIGA, '2115920'),
+          path: [],
+          slippage: 500,
+        },
+        options: { account: '0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa' },
+      },
+    ];
+
+    testCases.forEach(({ chainId, fields, options }) => {
+      it(`${fields.input.token.symbol} to ${fields.output.token.symbol}`, async function () {
+        const logic = new SwapTokenLogic(chainId);
+        const routerLogic = await logic.build(fields, options);
+        const sig = routerLogic.data.substring(0, 10);
+        const { tradeType, input, output, balanceBps } = fields;
+        const magicRouterAddress = getContractAddress(chainId, 'Router');
+
+        expect(routerLogic.to).to.eq(magicRouterAddress);
+        expect(utils.isBytesLike(routerLogic.data)).to.be.true;
+
+        if (tradeType === core.TradeType.exactIn) {
+          expect(sig).to.eq(iface.getSighash('swapExactTokensForTokens'));
+        } else {
+          expect(sig).to.eq(iface.getSighash('swapTokensForExactTokens'));
+        }
+
+        if (input.token.isNative) {
+          expect(routerLogic.inputs[0].token).to.eq(input.token.wrapped.address);
+        }
+
+        if (balanceBps) {
+          expect(routerLogic.inputs[0].balanceBps).to.eq(balanceBps);
+          expect(routerLogic.inputs[0].amountOrOffset).to.eq(common.getParamOffset(0));
+        } else {
+          expect(routerLogic.inputs[0].balanceBps).to.eq(core.BPS_NOT_USED);
+          expect(routerLogic.inputs[0].amountOrOffset).eq(input.amountWei);
+        }
+
+        expect(routerLogic.wrapMode).to.eq(
+          input.token.isNative
+            ? core.WrapMode.wrapBefore
+            : output.token.isNative
+            ? core.WrapMode.unwrapAfter
+            : core.WrapMode.none
+        );
+        expect(routerLogic.approveTo).to.eq(constants.AddressZero);
+        expect(routerLogic.callback).to.eq(constants.AddressZero);
```

Comment:
by the chance, could you explain what `approveTo` and `callback` are for?

Reply:
`approveTo` is for protocols that the entry contract is not the same as the token handling contract.
`callback` is for protocolink to be called again by the callback contract in a batch.